cudaaligner/src/myers_gpu.cu

```diff
@@ -586,27 +621,27 @@ __device__ void myers_compute_scores_diagonal_band_impl(
                     }
                 }
 
+                const WordType eq = get_query_pattern(query_patterns, idx, pattern_idx_offset + t - t_begin + 1, target_begin[t - 1], false);
+
                 const WordType carry_right_bit = WordType(1) << (idx == (n_words_band - 1) ? band_width - (n_words_band - 1) * word_size - 2 : word_size - 2);
                 const WordType carry_down_bit  = carry_right_bit << 1;
                 assert(carry_down_bit != 0);
-
-                const WordType eq = get_query_pattern(query_patterns, idx, pattern_idx_offset + t - t_begin + 1, target_begin[t - 1], false);
                 if (idx == n_words_band - 1)
                 {
                     // bits who have no left neighbor -> assume worst case: +1
                     pv_local |= carry_down_bit;
                     mv_local &= ~carry_down_bit;
                 }
 
-                const int32_t carry_right = myers_advance_block(warp_mask, carry_right_bit, eq, pv_local, mv_local, carry_down);
-                carry_down                = ((pv_local & carry_down_bit) == WordType(0) ? 0 : 1) - ((mv_local & carry_down_bit) == WordType(0) ? 0 : 1);
-                score(idx, t)             = score(idx, t - 1) + carry_right + carry_down;
+                const int2 carry_right   = myers_advance_block2(warp_mask, carry_right_bit, eq, pv_local, mv_local, carry);
+                const int32_t carry_down = ((pv_local & carry_down_bit) == WordType(0) ? 0 : 1) - ((mv_local & carry_down_bit) == WordType(0) ? 0 : 1);
+                score(idx, t)            = score(idx, t - 1) + carry_right.x + carry_down;
```

Comment:
I'm not intimately familiar with the details here, but flagging this just in case & I see that carry down is paired with carry_rught.x (as opposed to .y). Just making sure that's the right member of the struct?

Reply:
I need two "carries" along the horizontal: `carry_right.x` and `carry_right.y`, which correspond to the horizontal change in rows n-1 and n of the NW matrix, respectively, when going from column `t-1` to `t`.

The first "carry" I need to update `score(idx,t)`:
Since the score matrix is not the full NW matrix, but only the band along the diagonal, `idx` in `score(idx,t)` is relative to the first element of the band in the column of the NW matrix.
With respect to the full NW matrix, score(idx,t) moves down one row for each column. This means to get from `score(idx,t-1)` to `score(idx,t)`, I need to move one right and then one down, which results in score(idx,t) = score(idx, t-1)  + (change from column t-1 -> t in row n-1 in NW matrix) + (change in column t from row n-1 -> n in NW matrix).
Now, `carry_right.x` = (change from column t-1 -> t in row n-1 in NW matrix).

The second carry I need to actually "carry" from one warp iteration to the next:
The carry propagated from one warp iteration to the next, however, is the change from column `t-1` -> `t` in row n (not n-1). which is `carry_right.y`, because the next iteration will continue in row n+1.

Maybe `delta` would be a better variable name, since only one of the `carry_*` variables is an actual carry.

Reply:
Got it, thanks for the explanation @ahehn-nv ! Looks good to me, resolving :).